cdap-app-fabric/src/main/java/co/cask/cdap/internal/app/runtime/schedule/AbstractSchedulerService.java

```diff
@@ -90,280 +94,150 @@ protected final void stopScheduler() {
+    Set<Schedule> timeSchedules = Sets.newHashSet();
+    Set<Schedule> streamSizeSchedules = Sets.newHashSet();
+    for (Schedule schedule : schedules) {
+      if (schedule instanceof TimeSchedule) {
+        timeSchedules.add(schedule);
+      } else if (schedule instanceof StreamSizeSchedule) {
+        streamSizeSchedules.add(schedule);
+      } else {
+        throw new IllegalStateException("Unhandled type of schedule: " + schedule.getClass());
+      }
+    }
+    if (!timeSchedules.isEmpty()) {
+      timeScheduler.schedule(programId, programType, timeSchedules);
+    }
+    if (!streamSizeSchedules.isEmpty()) {
+      streamSizeScheduler.schedule(programId, programType, streamSizeSchedules);
+    }
   }
 
   @Override
   public List<ScheduledRuntime> nextScheduledRuntime(Id.Program program, SchedulableProgramType programType) {
-   return delegate.nextScheduledRuntime(program, programType);
+   return timeScheduler.nextScheduledRuntime(program, programType);
   }
 
   @Override
   public List<String> getScheduleIds(Id.Program program, SchedulableProgramType programType) {
-    return delegate.getScheduleIds(program, programType);
+    return ImmutableList.<String>builder()
+      .addAll(timeScheduler.getScheduleIds(program, programType))
+      .addAll(streamSizeScheduler.getScheduleIds(program, programType))
+      .build();
   }
 
   @Override
   public void suspendSchedule(Id.Program program, SchedulableProgramType programType, String scheduleName) {
-    delegate.suspendSchedule(program, programType, scheduleName);
+    ScheduleType type = getScheduleType(program, programType, scheduleName);
+    if (type == null) {
+      return;
+    }
+    switch (type) {
+      case TIME:
+        timeScheduler.suspendSchedule(program, programType, scheduleName);
+        break;
+      case STREAM:
+        streamSizeScheduler.suspendSchedule(program, programType, scheduleName);
+        break;
+    }
   }
 
   @Override
   public void resumeSchedule(Id.Program program, SchedulableProgramType programType, String scheduleName) {
-    delegate.resumeSchedule(program, programType, scheduleName);
+    ScheduleType type = getScheduleType(program, programType, scheduleName);
+    if (type == null) {
+      return;
+    }
+    switch (type) {
+      case TIME:
+        timeScheduler.resumeSchedule(program, programType, scheduleName);
+        break;
+      case STREAM:
+        streamSizeScheduler.resumeSchedule(program, programType, scheduleName);
+        break;
+    }
   }
 
   @Override
   public void deleteSchedule(Id.Program program, SchedulableProgramType programType, String scheduleName) {
-    delegate.deleteSchedule(program, programType, scheduleName);
+    ScheduleType type = getScheduleType(program, programType, scheduleName);
+    if (type == null) {
+      return;
+    }
+    switch (type) {
+      case TIME:
+        timeScheduler.deleteSchedule(program, programType, scheduleName);
```

Comment:
a general question about the scheduler methods: they all take the program name, not the Id.Program. So is the scheduler not yet namespace aware? I am pretty sure it will have to be... so perhaps it is better to pass down the Id.Program and not just the name...